### DISCUSSION TITLE
[@types/react-transition-group] RefHandler args in wrong order

### DISCUSSION THREAD

**Discussion by justintemps** [DefinitelyTyped/DefinitelyTyped#59684] [Issues with a @types package]
I may be wrong, but it seems like the arguments passed to the `RefHandler` generic are out of order:

```js
type RefHandler<
    RefElement extends undefined | HTMLElement,
    ImplicitRefHandler extends (node: HTMLElement, ...args: any[]) => void,
    ExplicitRefHandler extends (...args: any[]) => void
> = {
    implicit: ImplicitRefHandler;
    explicit: ExplicitRefHandler;
}[RefElement extends undefined ? 'implicit' : 'explicit'];

export type EndHandler<RefElement extends undefined | HTMLElement> = RefHandler<
    RefElement,
    (node: HTMLElement, done: () => void) => void,
    (done: () => void) => void
>;
```

Here we're saying that if there's no ref element, assume the implicit handler, if there is assume the explicit handler. Except that the implicit handler is expecting an HTMLelement node and the explicit handler isn't. 

Shouldn't it look like this insitead?

```js
type RefHandler<
    RefElement extends undefined | HTMLElement,
    ExplicitRefHandler extends (node: HTMLElement, ...args: any[]) => void,
    ImplicitRefHandler extends (...args: any[]) => void
> = {
    implicit: ImplicitRefHandler;
    explicit: ExplicitRefHandler;
}[RefElement extends undefined ? 'implicit' : 'explicit'];
```

[a comment by an automated account is not shown]

**Comment by MysteryBlokHed** [comment]
If both types `ImplicitRefHandler` and `ExplicitRefHandler` are mandatory, it shouldn't matter the order they are on the type. The check for which one is used is done here:

```typescript
// Gets the key 'implicit' or 'explicit' based on RefElement's type
[RefElement extends undefined ? 'implicit' : 'explicit']
```

If the `RefElement` type is `undefined`, then `ImplicitRefHandler` is used. If it is defined, then `ExplicitRefHandler` is used.

**Reply by justintemps** [reply]
Thanks @MysteryBlokHed. You're right, sorry, the order's not to do with it. But it still seems like `ImplicitRefHandler` and `ExplicitRefHandler`are mixed up. If `RefElement` is `undefined` then shouldn't the handler look `ImplicitRefHandler` look like this?

```js
 ImplicitRefHandler extends (...args: any[]) => void
```

Apologies again if I'm misunderstanding something.

**Reply by MysteryBlokHed** [reply]
Sorry, I misread your original question. I'm not familiar enough with the types to give an answer.
